frontend/src/employee-frontend/components/unit/unit-reservations/StaffAttendanceExternalPersonModal.tsx

```diff
@@ -49,18 +60,58 @@ type ExternalPersonModalProps = {
   defaultGroupId: UUID
 }
 
-const externalPersonForm = object({
-  date: validated(required(localDate()), (value) =>
-    value.isAfter(LocalDate.todayInHelsinkiTz()) ? 'dateTooLate' : undefined
-  ),
-  arrivalTime: required(localTime()),
-  departureTime: localTime(),
-  name: validated(string(), (value) =>
-    value.length < 3 ? 'required' : undefined
-  ),
-  group: required(oneOf<UUID>()),
-  hasStaffOccupancyEffect: required(boolean())
-})
+const externalPersonForm = transformed(
+  object({
+    date: validated(required(localDate()), (value) =>
+      value.isAfter(LocalDate.todayInHelsinkiTz()) ? 'dateTooLate' : undefined
+    ),
+    arrivalTime: required(localTime()),
+    departureTime: localTime(),
+    name: validated(string(), (value) =>
+      value.length < 3 ? 'required' : undefined
+    ),
+    group: required(oneOf<UUID>()),
+    hasStaffOccupancyEffect: required(boolean())
+  }),
+  ({
+    date,
+    arrivalTime,
+    departureTime,
+    name,
+    group,
+    hasStaffOccupancyEffect
+  }) => {
+    if (date.isToday() && arrivalTime.isAfter(LocalTime.nowInHelsinkiTz())) {
+      return ValidationError.field('arrivalTime', 'futureTime')
+    }
+    if (
+      date.isToday() &&
+      departureTime &&
+      departureTime.isAfter(LocalTime.nowInHelsinkiTz())
+    ) {
+      return ValidationError.field('departureTime', 'futureTime')
+    }
+    if (departureTime && !departureTime.isAfter(arrivalTime)) {
+      return ValidationError.of('timeRangeNotLinear')
+    }
```

Comment:
`TimeInterval` and its range operations are already available in the frontend and could be used here, too.

Reply:
True, but adding a `tryCreate` function that's already there for some other range types could hide range validation logic.

Reply:
https://github.com/espoon-voltti/evaka/pull/4803/commits/bebae94df8fa8ab28c9d5f15ba6e5b530d1193b1